### PR TITLE
add raw_ref macros

### PR DESCRIPTION
**Pull request by RalfJung**
In https://github.com/rust-lang/rust/issues/64490, various people were in favor of exposing `&raw` as a macro first before making the actual syntax stable. So this PR (unstably) introduces those macros. 

I'll create the tracking issue if we're okay moving forward with this.

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
This was discussed again in the @rust-lang/lang team meeting, and the general consensus in https://github.com/rust-lang/rust/issues/64490#issuecomment-627993209 has not changed. The question of whether some form of rust-lang/rfcs#2919 before *stabilizing* these macros was briefly discussed but not conclusively answered.

AFAIK, all this needs is sign-off from the T-libs on the precise syntax and the implementation.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

# Proposal

Introduce two macros

* `raw_ref!(path)`
* `raw_ref_mut!(path)`

which borrow the contents at `path` and produce (respectively) `*const T` and `*mut T` raw pointers. 

# Motivation for exposing these capabilities

Unlike the `&` syntax, no "safe reference" is created. This means that the resulting raw pointer can be misaligned
or invalid to dereference and that is not UB until the pointer is actually dereferenced. For more details see [RFC 2582].

[RFC 2582]: https://github.com/rust-lang/rfcs/pull/2582

# Motivation for using macro-based syntax

This is an alternative to the previous syntax specified in [RFC 2582], which was `&raw const` and `&raw mut`.

The motivations for making this change is that the syntax `&raw const` and `&raw mut` is not seen as particularly ideal. Among its downsides:

* It is verbose and wordy.
* It introduces a fresh contextual keyword `raw`.
* It breaks the symmetry that we previously have had, where `&foo` creates a `&T` type (and `&mut foo` creates a `&mut T`) type.

Further, we think that there may be an opportunity to revisit raw-pointer ergonomics in a bigger way that could improve the language overall (some background available in [these](https://github.com/rust-lang/rust/issues/27060#issuecomment-614745027) [three](https://github.com/rust-lang/rust/issues/27060#issuecomment-616567717) [messages](https://github.com/rust-lang/rust/issues/27060#issuecomment-616571760), particularly the final one). We don't feel ready to decide if that opportunity is worth the resulting churn, but we don't feel the need to commit to syntax like `&raw const` that may become deprecated.

# Future possibilities

The `raw_ref` and `raw_ref_mut` macros would be stable and hence usable in perpetuity. If we do wind up adopting a larger "raw pointer reform" than these macros might be deprecated in favor of the newer types and operators.

On the other hand, if we wind up deciding that larger "raw pointer reform" doesn't make sense, we might add some more "builtin" syntax than the `raw_ref` macros. Alternatively, we might simply keep the macros as they are, as they address a somewhat niche concern (creating raw pointers to invalid or unaligned memory, where the rules for a standard reference would be invalid).

**Comment by nikomatsakis** [comment]
Nominating for next lang-team meeting to draw attention these checkboxes.

**Comment by scottmcm** [comment]
@rfcbot reviewed

Definitely no concerns with this as unstable.

**Comment by mystor** [comment]
Due to the const provenance issues @Gankra mentioned in https://github.com/rust-lang/rust/issues/64490#issuecomment-626843521, I would think that we'd want people to prefer using `*mut` over `*const` when they can, as it probably produces the expected behaviour wrt stacked borrows. If that's still the case, it seems a bit unfortunate that the proposed macro names make the `mut` variant less ergonomic to write then the `const` variant, encouraging the use of `raw_ref!` by default.

Perhaps `raw_const_ref!` and `raw_mut_ref!` (or `raw_ref_const!`/`raw_ref_mut!`) would be better names?

**Comment by brain0** [comment]
> Perhaps `raw_const_ref!` and `raw_mut_ref!` (or `raw_ref_const!`/`raw_ref_mut!`) would be better names?

The usual Rust naming convention would be `raw_ref!` and `raw_mut!` (same as the `as_raw()` and `as_mut()` methods).

Another alternative would be `raw_const!` and `raw_mut!` to be consistent with `*const` and `*mut`. Since the result is not actually a reference, but a *raw pointer*, I would personally prefer this syntax.

**Comment by RalfJung** [comment]
I can change the name to anything as long as there's consensus. :) Don't really care myself.

**Comment by withoutboats** [comment]
maybe even raw_const_ptr / raw_mut_ptr. These are exposed from the root of std, right?

**Comment by RalfJung** [comment]
> These are exposed from the root of std, right?

Yes, but I can also make them "macro 2.0" macros and put them into some module -- I was not sure if that is already acceptable to do.

**Comment by nikomatsakis** [comment]
I don't have a strong opinion about the names we use, and I would love @rust-lang/libs feedback on the best way to expose them. I think something like `std::ptr::raw!(...)` or something might be nice, if we are doing that.

**Comment by sfackler** [comment]
I'd definitely prefer to have them be located in `ptr` rather than the crate root personally.

**Comment by withoutboats** [comment]
I think having them available only at the paths `std::ptr::raw_const` and `std::ptr::raw_mut` would be ideal, but I think this is the first time std would expose macros that way, and I'm not sure what the bigger picture of doing that is (does it work in 2015 edition / do we care? are we comfortable starting to do this with std macros in general? etc)

**Comment by dtolnay** [comment]
`raw_const!` and `raw_mut!` under core::ptr and std::ptr seem fine.

Alternatively, if it's a macro, do we need two? It could be `raw!(&const self.field)` & `raw!(&mut self.field)` where the macro accepts `&const $e:expr` and `&mut $e:expr`.

**Comment by nikomatsakis** [comment]
I believe @petrochenkov suggested that, from a compatibility POV, we could do that start to export macros from specific paths, but they'd be best positioned to weigh in on that.

**Comment by RalfJung** [comment]
Okay, so how do we resolve the naming bikeshed here? Some of the proposals on the table so far are
* `raw_ref!(path)`/`raw_ref_mut!(path)`
* `raw_const_ref!(path)`/`raw_mut_ref!(path)`
* `raw_const_ptr!(path)`/`raw_mut_ptr!(path)`
* `core::ptr::raw_const!(path)`/`core::ptr::raw_mut!(path)`
* `core::ptr::raw!(&const path)`/`core::ptr::raw!(&mut path)`

My main concern with the latter is that it looks like it is taking a normal mutable reference, and then doing something with it -- while it is absolutely crucial that this is *not* what happens.

**Comment by RalfJung** [comment]
I now implemented the second-to-last option: `core::ptr::raw_const!(path)`/`core::ptr::raw_mut!(path)`.

However, due to https://github.com/rust-lang/rust/issues/73286 I don't know if the doc tests are actually passing.

**Comment by petrochenkov** [comment]
In theory these macros should use `#[rustc_macro_transparency = "semitransparent"]` for full compatibility with `macro_rules`, but in practice it shouldn't matter what hygiene tokens `&`, `raw` and `mut` get.
(I'd still use just in case.)

**Comment by nikomatsakis** [comment]
@RalfJung 

> My main concern with the latter is that it looks like it is taking a normal mutable reference, and then doing something with it -- while it is absolutely crucial that this is not what happens.

Yeah, it is a bit "surprising" that what *looks* like an ordinary expression very much is not. On the other hand, the `&` makes it clear that the path here is "being borrowed" (i.e., we are taking its address, not evaluating it).

> I now implemented the second-to-last option: `core::ptr::raw_const!(path)/core::ptr::raw_mut!(path)`.

I think this is fine. I there is clear consensus that the macro should not be at the root, so +1 to `core::ptr::foo!`. Beyond that I guess there's not a clear winner.

**Comment by RalfJung** [comment]
CI failure is a network issue; I think this is ready for review.

File: src/libcore/ptr/mod.rs [hunk 1]
```diff
@@ -1399,3 +1399,70 @@ fnptr_impls_args! { A, B, C, D, E, F, G, H, I }
...
+///     f1: u8,
+///     f2: u16,
+/// }
+///
+/// let packed = Packed { f1: 1, f2: 2 };
+/// // `&packed.f2` would create an unaligned reference, and thus be Undefined Behavior!
+/// let raw_f2 = ptr::raw_const!(packed.f2);
+/// assert_eq!(unsafe { raw_f2.read_unaligned() }, 2);
+/// ```
+#[unstable(feature = "raw_ref_macros", issue = "none")]
```
Comment (nikomatsakis) [inline src/libcore/ptr/mod.rs:1429]: can we link these to some issue? perhaps the tracking issue https://github.com/rust-lang/rust/issues/64490 ? Or would you rather make a specialized tracking issue?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by joshtriplett** [comment]
:+1: for putting these in `core::ptr`. And I agree with the names that have been implemented: `raw_const!` and `raw_mut!` nicely parallel `*const` and `*mut`.

File: src/libcore/ptr/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/ptr/mod.rs:1429]: It's a different feature gate so it should probably be a separate tracking issue... I'll file one.

Comment (RalfJung) [reply src/libcore/ptr/mod.rs]: Tracking issue filed: https://github.com/rust-lang/rust/issues/73394

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**